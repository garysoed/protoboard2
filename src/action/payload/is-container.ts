import { StateId } from 'gs-tools/export/state';

export interface IsContainer {
  readonly $contentIds: StateId<readonly string[]>;
}
