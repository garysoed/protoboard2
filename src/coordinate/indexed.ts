export interface Indexed {
  readonly type: 'indexed';
  readonly index: number;
}

export function createIndexed(index: number): Indexed {
  return {type: 'indexed', index};
}
