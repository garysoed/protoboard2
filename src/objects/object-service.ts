import {source, stream} from 'grapevine';
import {$asMap, $filterNonNull, $map, $pipe} from 'gs-tools/export/collect';
import {StateId} from 'gs-tools/export/state';
import {NodeWithId, PersonaContext} from 'persona';
import {combineLatest, Observable, of as observableOf} from 'rxjs';
import {map, shareReplay} from 'rxjs/operators';

import {$getObjectSpec, $objectSpecIds} from './getters/root-state';
import {ObjectCreateSpec} from './object-create-spec';
import {ObjectSpec} from './object-spec';


class ObjectCache {
  private object: Observable<NodeWithId<Element>|null>|null = null;

  constructor(
      private readonly fn: ObjectCreateSpec<object>,
      readonly objectId: StateId<ObjectSpec<object>>,
  ) { }

  getOrCreate(context: PersonaContext): Observable<NodeWithId<Element>|null> {
    if (this.object) {
      return this.object;
    }

    const object = this.fn(this.objectId, context)
        .pipe(shareReplay({bufferSize: 1, refCount: true}));
    this.object = object;
    return object;
  }
}

// TODO: Move to RootState
export const $createSpecMap = source<ReadonlyMap<string, ObjectCreateSpec<any>>>(
    'createSpecMap',
    () => new Map(),
);

const $objectNodeCacheMap = stream<ReadonlyMap<string, ObjectCache>>(
    'objectNodeCacheMap',
    vine => combineLatest([
      $objectSpecIds.get(vine),
      $getObjectSpec.get(vine),
      $createSpecMap.get(vine),
    ])
        .pipe(
            map(([objectSpecIds, getObjectSpec, createSpecMap]) => {
              return $pipe(
                  objectSpecIds,
                  $map(id => {
                    const state = getObjectSpec(id);
                    if (!state) {
                      return null;
                    }

                    const handler = createSpecMap.get(state.type);
                    if (!handler) {
                      return null;
                    }

                    return [id.id, new ObjectCache(handler, id)] as const;
                  }),
                  $filterNonNull(),
                  $asMap(),
              );
            }),
            shareReplay({bufferSize: 1, refCount: true}),
        ),
);

type GetObjectNode = (id: StateId<ObjectSpec<any>>, context: PersonaContext) => Observable<NodeWithId<Element>|null>;
export const $getObjectNode = stream<GetObjectNode>(
    'getObjectNode',
    vine => $objectNodeCacheMap.get(vine).pipe(
        map(objectNodeCacheMap => {
          return (id: StateId<ObjectSpec<any>>, context: PersonaContext) => {
            const cache = objectNodeCacheMap.get(id.id);
            if (!cache) {
              return observableOf(null);
            }
            return cache.getOrCreate(context);
          };
        }),
    ),
);
