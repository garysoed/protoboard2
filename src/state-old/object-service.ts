import { source, Vine } from 'grapevine';
import { $asMap, $asSet, $filterNonNull, $map, $pipe } from 'gs-tools/export/collect';
import { cache } from 'gs-tools/export/data';
import { PersonaContext } from 'persona';
import { combineLatest, Observable, of as observableOf } from 'rxjs';
import { map, shareReplay, switchMap } from 'rxjs/operators';

import { ObjectCreateSpec } from './object-create-spec';
import { ObjectSpec } from './object-spec';
import { $objectSpecMap } from './object-spec-list';


class ObjectCache {
  private object: Observable<Node>|null = null;

  constructor(
      private readonly fn: ObjectCreateSpec<object>,
      private readonly state: ObjectSpec<object>,
  ) { }

  getOrCreate(context: PersonaContext): Observable<Node> {
    if (this.object) {
      return this.object;
    }

    const object = this.fn(this.state, context).pipe(shareReplay({bufferSize: 1, refCount: true}));
    this.object = object;
    return object;
  }
}


class ObjectService {
  constructor(private readonly vine: Vine) { }

  getObject(id: string, context: PersonaContext): Observable<Node|null> {
    return this.objectCachesMap$.pipe(
        switchMap(objectCachesMap => {
          const cache = objectCachesMap.get(id);
          if (!cache) {
            return observableOf(null);
          }

          return cache.getOrCreate(context);
        }),
    );
  }

  @cache()
  private get objectCachesMap$(): Observable<ReadonlyMap<string, ObjectCache>> {
    return combineLatest([$stateHandlers.get(this.vine), $objectSpecMap.get(this.vine)]).pipe(
        map(([stateHandlers, statesMap]) => {
          return $pipe(
              statesMap,
              $map(([id, state]) => {
                const handler = stateHandlers.get(state.type);
                if (!handler) {
                  return null;
                }

                return [id, new ObjectCache(handler, state)] as [string, ObjectCache];
              }),
              $filterNonNull(),
              $asMap(),
          );
        }),
        shareReplay({bufferSize: 1, refCount: true}),
    );
  }

  @cache()
  get objectIds$(): Observable<ReadonlySet<string>> {
    return this.objectCachesMap$.pipe(
        map(objectCachesMap => $pipe(objectCachesMap, $map(([id]) => id), $asSet())),
    );
  }
}

export const $objectService = source('ObjectService', vine => new ObjectService(vine));

export const $stateHandlers =
    source('stateHandlers', () => new Map<string, ObjectCreateSpec<object>>());
export function registerObjectCreateSpec<P extends object>(
    type: string,
    handler: ObjectCreateSpec<P>,
    vine: Vine,
): void {
  $stateHandlers.set(
      vine,
      existingHandlers => new Map([
        ...existingHandlers,
        [type, handler as ObjectCreateSpec<object>],
      ]),
  );
}
