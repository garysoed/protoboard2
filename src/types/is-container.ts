import {MutableState} from 'gs-tools/export/state';

import {ComponentId} from '../id/component-id';

export interface IsContainer {
  readonly contentIds: MutableState<ReadonlyArray<ComponentId<unknown>>>;
}