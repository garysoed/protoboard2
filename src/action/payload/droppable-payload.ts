import { IsContainer } from './is-container';

// TODO Delete
export type DroppablePayload = IsContainer;
