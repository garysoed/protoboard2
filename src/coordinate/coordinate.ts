import {Cartesian} from './cartesian';
import {Indexed} from './indexed';

export type Coordinate = Cartesian|Indexed;
