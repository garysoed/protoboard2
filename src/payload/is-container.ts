import {StateId} from 'gs-tools/export/state';

import {Indexed} from '../coordinate/indexed';
import {ObjectSpec} from '../types/object-spec';


export interface TypeCoordinateMapping {
  readonly indexed: Indexed;
}

export interface ContentSpec<C extends CoordinateTypes> {
  readonly objectId: StateId<ObjectSpec<any>>;
  readonly coordinate: TypeCoordinateMapping[C];
}

export type CoordinateTypes = keyof TypeCoordinateMapping;

export interface IsContainer<T extends CoordinateTypes> {
  readonly containerType: T;
  readonly $contentSpecs: StateId<ReadonlyArray<ContentSpec<T>>>;
}

interface IndexedSpec {
  readonly objectId: StateId<ObjectSpec<any>>;
  readonly coordinate: Indexed;
}
export function indexedContentSpec(spec: IndexedSpec): ContentSpec<'indexed'> {
  return spec;
}
