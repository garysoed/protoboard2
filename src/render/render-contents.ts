import { $asArray, $map, $pipe, $sort, normal, withMap } from 'gs-tools/export/collect';
import { filterNonNull } from 'gs-tools/export/rxjs';
import { $stateService } from 'mask';
import { PersonaContext } from 'persona';
import { MultiOutput } from 'persona/export/internal';
import { combineLatest, Observable, of as observableOf } from 'rxjs';
import { switchMap, withLatestFrom } from 'rxjs/operators';

import { Indexed } from '../coordinate/indexed';
import { $objectService } from '../objects/object-service';
import { IsContainer } from '../payload/is-container';


export function renderContents(
    isContainer$: Observable<IsContainer<'indexed'>|null>,
    output: MultiOutput,
    context: PersonaContext,
): Observable<unknown> {
  return combineLatest([$stateService.get(context.vine), isContainer$]).pipe(
      switchMap(([stateService, isContainer]) => {
        if (!isContainer) {
          return observableOf(null);
        }

        return stateService.get(isContainer.$contentSpecs);
      }),
      withLatestFrom($objectService.get(context.vine)),
      switchMap(([contentIds, renderableService]) => {
        const node$list = $pipe(
            contentIds ?? [],
            $sort(withMap(({coordinate}) => coordinate.index, normal())),
            $map(({objectId}) => {
              return renderableService.getObject(objectId, context).pipe(filterNonNull());
            }),
            $asArray(),
        );

        return node$list.length <= 0 ? observableOf([]) : combineLatest(node$list);
      }),
      output.output(context),
  );
}

function renderIndexed(contents: ReadonlyMap<Indexed, Node>): readonly Node[] {
  return $pipe(
      [...contents],
      $sort(withMap(([coordinate]) => coordinate.index, normal())),
      $map(([, node]) => node),
      $asArray(),
  );
}

