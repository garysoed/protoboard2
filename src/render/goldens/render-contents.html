<pbt-test><gst-fake-shadow><div id="container"><!-- #ref -->onetwothree</div></gst-fake-shadow></pbt-test>